m3u8/writer.go

```diff
@@ -966,8 +969,10 @@ func (p *MediaPlaylist) encode(segmentsToSkipInTotal uint64) *bytes.Buffer {
 			writeDateRange(&p.buf, seg.SCTE35DateRanges[i], p.WritePrecision())
 		}
 		// check for key change
-		if seg.Key != nil && (p.Key == nil || *seg.Key != *p.Key) {
-			writeKey("#EXT-X-KEY:", &p.buf, seg.Key)
+		if len(seg.Keys) != 0 && (p.Keys == nil || !reflect.DeepEqual(seg.Keys, p.Keys)) {
```

Comment:
I think one should try to avoid using the reflect.DeepEqual outside of tests.

`Key` is a comparable struct, since it only contains comparable fields.

The only thing to do then is to compare two slices of keys.
We can do this either by implementing our own function, or changing the minimal Go version to 1.21 and use the generic slices package like `slices.Equal(seg.Keys, p.Keys)`. I think I prefer the later.